pdg/src/info.rs

```diff
@@ -67,18 +67,486 @@ fn set_flow_info(g: &mut Graph) {
         }
         node.info = Some(NodeInfo {
             flows_to: cur_node_flow_info,
+            unique: false,
         });
     }
 }
 
+/// Gathers information from a [`Graph`] (assumed to be acyclic and topologically sorted but not
+/// necessarily connected) for each [`Node`] in it what its chronologically (judged by [`NodeId`])
+/// final descendent is.
+fn get_last_desc(g: &mut Graph) -> HashMap<NodeId, NodeId> {
+    let mut desc_map: HashMap<NodeId, NodeId> =
+        HashMap::from_iter(g.nodes.iter_enumerated().map(|(idx, _)| (idx, idx)));
+    for (n_id, node) in g.nodes.iter_enumerated().rev() {
+        if let Some(p_id) = node.source {
+            let cur_node_last_desc: NodeId = *desc_map.get(&n_id).unwrap();
+            let parent_last_desc: NodeId = desc_map.remove(&p_id).unwrap();
+            desc_map.insert(p_id, std::cmp::max(cur_node_last_desc, parent_last_desc));
+        }
+    }
+    desc_map
+}
+
+/// Finds the inverse of a [`Graph`], each [`Node`] mapping to a list of its children.
+fn collect_children(g: &Graph) -> HashMap<NodeId, Vec<NodeId>> {
+    let mut m = HashMap::new();
+    for (par, chi) in g
+        .nodes
+        .iter_enumerated()
+        .filter_map(|(idx, node)| Some((node.source?, idx)))
+    {
+        m.entry(par).or_insert_with(Vec::new).push(chi)
+    }
+    for par in g.nodes.indices() {
+        let _ = m.try_insert(par, Vec::new());
+    }
+    m
+}
+
+///Given a list of nodes of the same parent and information about them,
+///determines if any have conflicts with any of the others.
+///Children which are not a field cannot be live at the same time as any other child.
+///Children which are a field cannot be live at the same time as any other one of the same field.
+fn check_children_conflict(
+    g: &Graph,
+    n_id: &NodeId,
+    children: &HashMap<NodeId, Vec<NodeId>>,
+    descs: &HashMap<NodeId, NodeId>,
+) -> bool {
+    let mut max_descs: HashMap<Option<Field>, NodeId> = HashMap::new();
+    for id in children.get(n_id).unwrap() {
+        let sib_node: &Node = g.nodes.get(*id).unwrap();
+        let my_last_desc = descs.get(&id).unwrap().clone();
+        print!("{}\n", my_last_desc);
+        if matches!(max_descs.get(&None), Some(max_desc) if max_desc > id)
+            || matches!(sib_node.kind,NodeKind::Field(f) if matches!(max_descs.get(&Some(f)),Some(max_desc_field) if max_desc_field > id))
```

Comment:
have you run `rustfmt`?

Reply:
This part isn't changed by it.

Reply:
i would add some spacing after the comments and see if that changes, makes things more readable IMO

Reply:
I've done that, but it hasn't added any newlines, if you want me to add one I'm not sure where to put it.

Reply:
it's fine, not the biggest deal